test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java

```diff
@@ -455,6 +460,23 @@ protected void ensureNoInitializingShards() {
         assertFalse("timed out waiting for shards to initialize", actionGet.isTimedOut());
     }
 
+    /**
+     * waits until all free_context actions have been handled by the generic thread pool
+     */
+    protected void ensureAllFreeContextActionsAreConsumed() throws Exception {
+        NodeService nodeService = getInstanceFromNode(NodeService.class);
+        MockSearchTransportService searchTransportService = (MockSearchTransportService) nodeService.getSearchTransportService();
+        assertBusy(() -> {
+            assertThat(
+                searchTransportService.getFreeContextRequests().get(),
+                equalTo(
+                    searchTransportService.getSuccessfulFreeContextRequests().get() + searchTransportService.getFailedFreeContextRequests()
+                        .get()
+                )
+            );
+        }, 5, TimeUnit.SECONDS);
```

Comment:
5 sec is probably too short, tests can run quite slowly sometimes. I'd recommend `SAFE_AWAIT_TIMEOUT` (10 sec) here.

Reply:
Will refactor as per Ben's suggestion below, which also makes use of the `SAFE_AWAIT_TIMEOUT` of 10sec :)